src/main/java/de/fearnixx/jeak/teamspeak/voice/connection/VoiceConnection.java

```diff
@@ -0,0 +1,223 @@
+        this.port = port;
+        this.eventService = eventService;
+        this.userService = userService;
+        this.bot = bot;
+        locked = true;
+    }
+
+    @Override
+    public void connect(Runnable onSuccess, Runnable onError) {
+        if (connected) {
+            return;
+        }
+
+        new Thread(
+                () -> {
+                    this.ts3jClientSocket = new LocalTeamspeakClientSocket();
+
+                    ts3jClientSocket.setNickname(clientConnectionInformation.getClientNickname());
+                    ts3jClientSocket.setIdentity(clientConnectionInformation.getTeamspeakIdentity());
+
+                    ts3jClientSocket.addListener(
+                            new TS3Listener() {
+                                @Override
+                                public void onTextMessage(TextMessageEvent e) {
+                                    eventService.fireEvent(
+                                            new VoiceConnectionTextMessageEvent(clientConnectionInformation.getIdentifier(), e)
+                                    );
+
+                                    if (shouldForwardTextMessages) {
+                                        final QueryEvent.ClientTextMessage clientTextMessage = new QueryEvent.ClientTextMessage(userService);
+
+                                        final int invokerId = e.getInvokerId();
+                                        final IClient client = userService.getClientByID(invokerId).orElseThrow();
+
+                                        clientTextMessage.setClient(client);
+                                        clientTextMessage.setCaption(EventCaptions.TEXT_MESSAGE);
+                                        clientTextMessage.setConnection(bot.getServer().getConnection());
+                                        clientTextMessage.setProperty(PropertyKeys.TextMessage.MESSAGE, e.getMessage());
+                                        clientTextMessage.setProperty("invokerid", invokerId);
+                                        clientTextMessage.setProperty("invokeruid", client.getClientUniqueID());
+                                        clientTextMessage.setProperty("invokername", client.getNickName());
+                                        clientTextMessage.setProperty("targetmode", e.getTargetMode());
+
+                                        eventService.fireEvent(clientTextMessage);
+                                    }
+                                }
+                            }
+                    );
+
+                    InetSocketAddress inetSocketAddress = new InetSocketAddress(hostname, port);
+                    try {
+                        ts3jClientSocket.connect(inetSocketAddress, null, 10000L);
+                    } catch (IOException | TimeoutException e) {
+                        onError.run();
+                        return;
+                    }
+
+                    this.connected = true;
```

Comment:
Due to the timeout possibly blocking the current thread, the whole method should be synchronized to prevent simultaneous calls to ``#connect`` from different threads.